virtcontainers/acrn.go

```diff
@@ -0,0 +1,630 @@
+			if err := os.RemoveAll(vmPath); err != nil {
+				a.Logger().WithError(err).Error("Fail to clean up vm directory")
+			}
+		}
+	}()
+
+	var strErr string
+	var PID int
+	PID, strErr, err = LaunchAcrn(a.acrnConfig, virtLog.WithField("subsystem", "acrn-dm"))
+	if err != nil {
+		return fmt.Errorf("%s", strErr)
+	}
+	a.info.PID = PID
+
+	if err = a.waitSandbox(timeout); err != nil {
+		a.Logger().WithField("Acrn wait failed:", err).Debug()
+		return err
+	}
+
+	//Store VMM information
+	return a.store.Store(store.Hypervisor, a.info)
+
+}
+
+// waitSandbox will wait for the Sandbox's VM to be up and running.
+func (a *acrn) waitSandbox(timeout int) error {
+	span, _ := a.trace("waitSandbox")
+	defer span.Finish()
+
+	if timeout < 0 {
+		return fmt.Errorf("Invalid timeout %ds", timeout)
+	}
+
+	time.Sleep(time.Duration(timeout) * time.Second)
+
+	return nil
+}
+
+// stopSandbox will stop the Sandbox's VM.
+func (a *acrn) stopSandbox() (err error) {
+
+	span, _ := a.trace("stopSandbox")
+	defer span.Finish()
+
+	a.Logger().Infof("Stopping ACRN VM")
+
+	defer func() {
+		if err != nil {
+			a.Logger().Infof("stopSandbox failed")
+		} else {
+			a.Logger().Infof("ACRN VM stopped")
+		}
+	}()
+
+	pid := a.info.PID
+
+	// Check if VM process is running, in case it is not, let's
+	// return from here.
+	if err = syscall.Kill(pid, syscall.Signal(0)); err != nil {
+		a.Logger().Infof("ACRN VM already stopped")
+		return nil
+	}
+
+	// Send a SIGTERM to the VM process to try to stop it properly
+	if err = syscall.Kill(pid, syscall.SIGINT); err != nil {
+		a.Logger().Infof("Sending SIGINT to stop ACRN VM failed")
+		return err
+	}
+
+	// Wait for the VM process to terminate
+	tInit := time.Now()
+	for {
+		if err = syscall.Kill(pid, syscall.Signal(0)); err != nil {
+			a.Logger().Infof("ACRN VM stopped after sending SIGINT")
+			return nil
+		}
+
+		if time.Since(tInit).Seconds() >= acrnStopSandboxTimeout {
+			a.Logger().Warnf("VM still running after waiting %ds", acrnStopSandboxTimeout)
+			break
+		}
+
+		// Let's avoid to run a too busy loop
+		time.Sleep(time.Duration(50) * time.Millisecond)
+	}
+
+	// Let's try with a hammer now, a SIGKILL should get rid of the
+	// VM process.
+	return syscall.Kill(pid, syscall.SIGKILL)
+
+}
+
+func (a *acrn) udpateBlockDevice(drive *config.BlockDrive) error {
+	var err error
+	if drive.File == "" || drive.Index >= AcrnBlkDevPoolSz {
+		return fmt.Errorf("Empty filepath or invalid drive index, Dive ID:%s, Drive Index:%d",
+			drive.ID, drive.Index)
+	}
+
+	slot := AcrnBlkdDevSlot[drive.Index]
+
+	//Explicitly set PCIAddr to NULL, so that VirtPath can be used
+	drive.PCIAddr = ""
+
+	args := []string{"blkrescan", a.acrnConfig.Name, fmt.Sprintf("%d,%s", slot, drive.File)}
+
+	a.Logger().WithFields(logrus.Fields{
+		"drive": drive,
+		"path":  a.config.HypervisorCtlPath,
+	}).Infof("udpateBlockDevice with acrnctl path")
+	cmd := exec.Command(a.config.HypervisorCtlPath, args...)
+	if err := cmd.Run(); err != nil {
```

Comment:
Can't this be simplified to just:

```go
return cmd.Run()
```

The caller should catch this and log the error?

Reply:
The caller of updateBlockDevice<-hotplugAddDevice<-HotplugAddDevice<-Attach<-AttachDevice. None of these seems to log the failure and this log will not be meaningful in caller as it is specific to updateBlockDevice.